be/src/olap/short_key_index.h

```diff
@@ -0,0 +1,183 @@
+    uint32_t _segment_id = 0;
+    uint32_t _num_rows_per_block;
+    bool _delete_flag;
+
+    // short key length
+    size_t _key_length;
+
+    // used to save serialized key temporarily
+    faststring _tmp_key_buf;
+
+    faststring _header_buf;
+    faststring _index_buf;
+
+    // NOTE: most of these two fields are useless.
+    // put here just for compatible. I will remove this one day
+    // only used to compatible with old version
+    FileHeader<OLAPIndexHeaderMessage, OLAPIndexFixedHeader> _header;
+};
+
+// Used to decode short key to header and encoded index data.
+// Actually, this class should decode index data to memory format.
+// But for compatible with old code, we reuse MemIndex class.
+// Usage:
+//      MemIndex index;
+//      ShortKeyIndexDecoder decoder(slice)
+//      decoder.parse();
+//      index.load_segment(decoder.header(), decoder.index_data());
+class ShortKeyIndexDecoder {
+public:
+    ShortKeyIndexDecoder(const Slice& data) : _data(data) { }
+    Status parse();
+    const FileHeader<OLAPIndexHeaderMessage, OLAPIndexFixedHeader>& header() const {
+        return _header;
+    }
+    const Slice& index_data() const { return _index_data; }
+private:
+    Slice _data;
+
+    // valid after parse() is called, Header has been removed.
+    // Only contains index data 
+    Slice _index_data;
+
+    // only used to compatible with old version
+    FileHeader<OLAPIndexHeaderMessage, OLAPIndexFixedHeader>  _header;
+};
+
+class ShortKeyIndexIterator {
+public:
+    ShortKeyIndexIterator(const MemIndex* index) : _index(index), _segment(0), _block(0) {
+        _helper.init(_index->short_key_fields());
+    }
+
+    // Position at the first key in the index that at or past target
+    inline void seek_to(const RowCursor& key) {
+        auto offset = _index->find(key, &_helper, false);
+        if (offset.offset > 0) {
+            offset.offset -= 1;
+            offset = _index->next(offset);
+        }
+        _segment = offset.segment;
+        _block = offset.offset;
+    }
+
+    // Position at the first key in the index that post target
+    inline void seek_after(const RowCursor& key) {
+        auto offset = _index->find(key, &_helper, true);
+        if (offset.offset > 0) {
+            offset.offset -= 1;
+            offset = _index->next(offset);
+        }
+        _segment = offset.segment;
+        _block = offset.offset;
+    }
+
+    // Position at the first key in the index that before target
+    inline void seek_before(const RowCursor& key) {
+        seek_to(key);
+        if (valid()) {
+            prev();
+        }
+    }
+
+    inline bool valid() const {
+        return _segment < _index->segment_count();
+    }
+
+    inline void prev() {
+        DCHECK(valid());
+        OLAPIndexOffset offset(_segment, _block);
+        auto new_offset = _index->prev(offset);
+        _segment = new_offset.segment;
+        _block = new_offset.offset;
+    }
+
+    inline void next() {
```

Comment:
missing doc, especially the expected behavior at corner case

Reply:
OK, I will add comment